### PR TITLE
feat: add Slider component

### PR DESCRIPTION
**Pull request by NadiaPosch**

### CODE REVIEW

File: src/components/Slider/Slider.tsx [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import { Dispatch, ReactElement, SetStateAction } from "react";
+import css from "./Slider.module.css";
+
+export interface IconItem {
+    icon: ReactElement<IconProps>;
+    name: string;
+}
+
+export interface SliderProps {
+    items: string[] | IconItem[];
```
Comment (dwirz) [inline src/components/Slider/Slider.tsx:13]: IMO it would be better to use an object with an `id` prop instead of plain string. atm if the string is translated the dev who uses the `Slider` somehow needs to map the selected string to a unique item outside of the `Slider`. do you know what i mean? 🙂

File: src/components/Slider/Slider.tsx [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+const isIconItem = (item: string | IconItem): item is IconItem => (item as IconItem).icon !== undefined;
+
+export default function Slider({ items, activeItem: active, onChange }: SliderProps): ReactElement<SliderProps> {
+    return (
+        <div data-test-id="Slider" className={css.wrapper}>
+            {items.map((item, i) =>
+                isIconItem(item) ? (
+                    <div
+                        key={i}
```
Comment (dwirz) [inline src/components/Slider/Slider.tsx:26]: if the item has an `id`-prop use this instead of `i`. also using the array index is discouraged to use if there is another unique value you can use. (also see comment on SO https://stackoverflow.com/a/46735689 why using index should be avoided)

File: src/components/Slider/Slider.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,32 @@
+/* (c) Copyright Frontify Ltd., all rights reserved. */
+
+import { mount } from "@cypress/react";
+import { useState } from "react";
+import Slider from "./Slider";
+import css from "./Slider.module.css";
+
+const Slider_Items = ["abc", "def", "ghi"];
```
Comment (dwirz) [inline src/components/Slider/Slider.spec.tsx:8]: not sure if this is a convention of frontify (@SamuelAlev) but @NadiaPosch IMO either write everything in capital letters eg `SLIDER_ITEMS` or classical snake_case. this is more of a gusto thingy but personally i never saw this kind of variable "convention" 😉

File: src/components/Slider/Slider.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+export interface SliderProps {
+    items: string[] | IconItem[];
+    activeItem: string;
+    onChange: Dispatch<SetStateAction<string>>;
+}
+
+const isIconItem = (item: string | IconItem): item is IconItem => (item as IconItem).icon !== undefined;
+
+export default function Slider({ items, activeItem: active, onChange }: SliderProps): ReactElement<SliderProps> {
```
Comment (dwirz) [inline src/components/Slider/Slider.tsx:20]: @SamuelAlev not sure if naming this component `Slider` is somehow misleading, wdyt? (if frontify always names these kind of component slider i'm fine with it 🙂 )

File: src/components/Slider/Slider.tsx (hunk 1 above)
Comment (NadiaPosch) [reply src/components/Slider/Slider.tsx:13]: yes, good idea!

File: .husky/pre-push [hunk 5]
```diff
@@ -1,4 +0,0 @@
-#!/bin/sh
-. "$(dirname "$0")/_/husky.sh"
-
-npm test
```
Comment (dwirz) [inline .husky/pre-push:4]: why did you remove the pre-push hook?

File: src/components/Slider/Slider.tsx [hunk 6]
```diff
@@ -5,37 +5,36 @@ import { Dispatch, ReactElement, SetStateAction } from "react";
...
 
-const isIconItem = (item: string | IconItem): item is IconItem => (item as IconItem).icon !== undefined;
+const isIconItem = (item: TextItem | IconItem): item is IconItem => (item as IconItem).icon !== undefined;
+
+export default function Slider({ items, activeItemId, onChange }: SliderProps): ReactElement<SliderProps> {
+    const getItemClasses = (id: string) => [css.item, id === activeItemId ? css.active : ""].join(" ");
 
-export default function Slider({ items, activeItem: active, onChange }: SliderProps): ReactElement<SliderProps> {
     return (
         <div data-test-id="Slider" className={css.wrapper}>
```
Comment (dwirz) [inline src/components/Slider/Slider.tsx:29]: actually since it is a "list" of items you could use the "semantically correct"  html element `<ul>` and items as `<li>` wdyt?

**Review by dwirz — CHANGES_REQUESTED** [review]

File: src/components/Slider/Slider.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+export interface SliderProps {
+    items: TextItem[] | IconItem[];
+    activeItemId: string;
+    onChange: Dispatch<SetStateAction<string>>;
+}
+
+const isIconItem = (item: TextItem | IconItem): item is IconItem => (item as IconItem).icon !== undefined;
+
+export default function Slider({ items, activeItemId, onChange }: SliderProps): ReactElement<SliderProps> {
+    const getItemClasses = (id: string) => [css.item, id === activeItemId ? css.active : ""].join(" ");
```
Comment (SamuelAlev) [inline src/components/Slider/Slider.tsx:26]: It can be moved out of the function

File: src/components/Slider/Slider.tsx [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+    return (
+        <ul data-test-id="Slider" className={css.wrapper}>
```
Comment (SamuelAlev) [inline src/components/Slider/Slider.tsx:29]:
```suggestion
        <ul data-test-id="slider" className={css.wrapper}>
```

We use kebab-case for the id

File: src/components/Slider/Slider.tsx [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+export interface TextItem {
+    id: string;
+    name: string;
+}
+
+export interface SliderProps {
+    items: TextItem[] | IconItem[];
+    activeItemId: string;
+    onChange: Dispatch<SetStateAction<string>>;
```
Comment (SamuelAlev) [inline src/components/Slider/Slider.tsx:20]:
```suggestion
    onChange: (id: string) => void;
```

We don't use Redux within the product

File: src/components/Slider/Slider.tsx [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+const isIconItem = (item: TextItem | IconItem): item is IconItem => (item as IconItem).icon !== undefined;
+
+export default function Slider({ items, activeItemId, onChange }: SliderProps): ReactElement<SliderProps> {
+    const getItemClasses = (id: string) => [css.item, id === activeItemId ? css.active : ""].join(" ");
+
+    return (
+        <ul data-test-id="Slider" className={css.wrapper}>
+            {items.map((item) =>
+                isIconItem(item) ? (
+                    <li key={item.id} onClick={() => onChange(item.id)} className={getItemClasses(item.id)}>
```
Comment (SamuelAlev) [inline src/components/Slider/Slider.tsx:32]: It is missing the actual selection change (clicking the second item doesn't put the active state on it)

File: src/components/Slider/Slider.stories.tsx [hunk 11]
```diff
@@ -0,0 +1,22 @@
+/* (c) Copyright Frontify Ltd., all rights reserved. */
```
Comment (SamuelAlev) [inline src/components/Slider/Slider.stories.tsx:1]: It misses

```ts
Default.argTypes = {
    onChange: { action: "onClear" },
};
```

File: src/components/Slider/Slider.stories.tsx [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+export default {
+    title: "Slider",
+    component: Slider,
+} as Meta;
+
+const SliderTemplate: Story<SliderProps> = (args) => <Slider {...args} />;
+
+export const Default = SliderTemplate.bind({});
+
+Default.args = {
```
Comment (SamuelAlev) [inline src/components/Slider/Slider.stories.tsx]: We could probably add another story with icons only

File: src/components/Slider/Slider.spec.tsx [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+describe("Slider Component", () => {
+    it("should render foo text correctly", () => {
+        const Component = () => {
+            const [active, setActive] = useState(SLIDER_ITEMS[0].id);
+
+            return <Slider items={SLIDER_ITEMS} activeItemId={active} onChange={setActive} />;
+        };
+
+        mount(<Component />);
```
Comment (SamuelAlev) [inline src/components/Slider/Slider.spec.tsx:22]:
```suggestion
        const onChangeStub = cy.stub().as("onChangeStub");

        mount(<Slider items={SLIDER_ITEMS} activeItemId={SLIDER_ITEMS[0].id} onChange={onChangeStub} />);
```

We can use cypress stub here

File: src/components/Slider/Slider.spec.tsx [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+import css from "./Slider.module.css";
+
+const SLIDER_ITEMS = [
+    { id: "a", name: "abc" },
+    { id: "b", name: "def" },
+    { id: "c", name: "ghi" },
+];
+
+describe("Slider Component", () => {
+    it("should render foo text correctly", () => {
```
Comment (SamuelAlev) [inline src/components/Slider/Slider.spec.tsx:15]:
```suggestion
    it("renders", () => {
```
I would move out the click to its own test `it('handles click and change the active item')`

**Review by SamuelAlev — CHANGES_REQUESTED** [review]
Nice one! Just got a few points to fix otherwise looks great

File: src/components/Slider/Slider.tsx (hunk 10 above)
Comment (NadiaPosch) [reply src/components/Slider/Slider.tsx:32]: The state change will be handled by the parent component so that the active item is also accessible from outside the Slider. I have updated the story to reflect a working example.

File: src/components/Slider/Slider.spec.tsx (hunk 13 above)
Comment (NadiaPosch) [reply src/components/Slider/Slider.spec.tsx:22]: I'm not sure if a stub would work for the state change. I have updated the tests, would be great if you could have another look at it :)

**Review by dwirz — APPROVED** [review]

File: src/components/Slider/Slider.tsx [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+    [css.item, id === activeItemId ? css.active : ""].join(" ");
+
+export default function Slider({ items, activeItemId, onChange }: SliderProps): ReactElement<SliderProps> {
+    return (
+        <ul data-test-id="slider" className={css.wrapper}>
+            {items.map((item) =>
+                isIconItem(item) ? (
+                    <li
+                        key={item.id}
+                        onClick={() => onChange(item.id)}
```
Comment (SamuelAlev) [inline src/components/Slider/Slider.tsx:34]: > The state change will be handled by the parent component so that the active item is also accessible from outside the Slider. I have updated the story to reflect a working example.

What are the pros of doing of moving out the active item id?
My way of doing it would be like:
```ts
const onItemClicked = (id: string) => {
	setActiveItemId(id) // Local useState
	onChange && onChange(id)
}

//[...]
<li
    key={item.id}
    onClick={() => onItemClicked(item.id)}
    className={getItemClasses(item.id, activeItemId)}
>
```

File: src/components/Slider/Slider.spec.tsx [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+};
+
+beforeEach(() => {
+    mount(<Component />);
+    cy.get("[data-test-id=slider]").as("slider");
+    cy.get("@slider").children().first().as("firstItem");
+    cy.get("@slider").children().last().as("lastItem");
+});
+
+describe("Slider Component", () => {
```
Comment (SamuelAlev) [inline src/components/Slider/Slider.spec.tsx:26]: We could also add an icons slider in the tests as the result may change depending on texts vs icons.

**Review by SamuelAlev — COMMENTED** [review]
Let's talk about my comments this Monday :)

File: .husky/pre-push (hunk 5 above)
Comment (NadiaPosch) [inline .husky/pre-push:4]: I'm not sure how to undo the deletion of the pre-push file. Could you help me with that @SamuelAlev ?

Comment (SamuelAlev) [reply .husky/pre-push:4]: `git checkout main .husky/pre-push` and then commit it back
